proof/refine/ARM/Syscall_R.thy

```diff
@@ -858,6 +858,171 @@ lemma threadSet_not_curthread_ct_domain:
   apply clarsimp
   done
 
+lemma schedContextBindNtfn_invs':
+  "\<lbrace>invs' and ex_nonz_cap_to' scPtr and ex_nonz_cap_to' ntfnPtr\<rbrace>
+   schedContextBindNtfn scPtr ntfnPtr
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  apply (clarsimp simp: schedContextBindNtfn_def)
+  apply (wpsimp wp: setSchedContext_invs' setNotification_invs' hoare_vcg_imp_lift'
+                    hoare_vcg_all_lift getNotification_wp)
+  apply (rule conjI)
+   apply (fastforce dest: ntfn_ko_at_valid_objs_valid_ntfn'
+                    simp: valid_ntfn'_def
+                   split: ntfn.splits)
+  apply (fastforce dest: sc_ko_at_valid_objs_valid_sc'
+                   simp: valid_sched_context'_def valid_sched_context_size'_def objBits_simps)
+  done
+
+lemma contextYieldToUpdateQueues_invs'_helper:
+   "\<lbrace>\<lambda>s. invs' s \<and> sc_at' scPtr s \<and> valid_sched_context' sc s \<and> valid_sched_context_size' sc
+         \<and> ex_nonz_cap_to' scPtr s \<and> ex_nonz_cap_to' ctPtr s \<and> tcb_at' ctPtr s\<rbrace>
+    do y \<leftarrow> threadSet (tcbYieldTo_update (\<lambda>_. Some scPtr)) ctPtr;
+       setSchedContext scPtr (scYieldFrom_update (\<lambda>_. Some ctPtr) sc)
+    od
+    \<lbrace>\<lambda>_. invs'\<rbrace>"
+   apply (clarsimp simp: invs'_def valid_pspace'_def valid_dom_schedule'_def)
+   apply (wp threadSet_valid_objs' threadSet_mdb' threadSet_iflive' threadSet_cap_to
+             threadSet_ifunsafe'T   threadSet_ctes_ofT threadSet_valid_queues_new
+             threadSet_valid_queues' threadSet_valid_release_queue threadSet_valid_release_queue'
+             untyped_ranges_zero_lift valid_irq_node_lift valid_irq_handlers_lift''
+             hoare_vcg_const_imp_lift hoare_vcg_imp_lift' threadSet_valid_replies'
+          | clarsimp simp: tcb_cte_cases_def cteCaps_of_def)+
+   apply (fastforce simp: obj_at_simps valid_tcb'_def tcb_cte_cases_def comp_def
+                          valid_sched_context'_def valid_sched_context_size'_def
+                          valid_release_queue'_def inQ_def)
+   done
+
+lemma setSchedContext_ct_in_state':
+  "\<lbrace>ct_in_state' P and sc_at' ptr\<rbrace> setSchedContext ptr sc \<lbrace>\<lambda>_. ct_in_state' P\<rbrace>"
+  apply (wpsimp wp: set_sc'.set_wp)
+  apply (fastforce simp: ct_in_state'_def st_tcb_at'_def obj_at_simps ps_clear_def)
+  done
```

Comment:
I don’t think we need `sc_at’` in the precondition. We already have a lemma about `setSchedContext` and `pred_tcb_at’`. That with `hoare_lift_Pf` or something like that should work.

Reply:
Done, and put in `SchedContext_R.thy`